File: bookOfProblems/leetcode.com/javascript/26. Remove Duplicates from Sorted Array.js
/**
 * @param {number[]} nums
 * @return {number}
 */

let nums = [1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 5, 6, 7]

// var removeDuplicates = function(nums) {
//     let index = 0;

//     for(let i=0; i<nums.length; i++){
//         if(nums[i] !== nums[i+1]){
//             nums[index++] = nums[i];
//         }
//     }
//     return index;
// };

var removeDuplicates = function (nums) {
    let set = new Set(nums);
    nums.length = 0;
    nums.push(...set.values())
    return set.size;
};

console.log(removeDuplicates(nums));

////////////////////////////////////////////////////////////////////////////////

// function removeDuplicates(nums: number[]): number {
//     for (let i = 1; i < nums.length; i++) {
//         if (nums[i] === nums[i - 1]) {
//             nums.splice(i, 1);
//             i--;
//         }
//     }

//     return nums.length;
// };

////////////////////////////////////////////////////////////////////////////////

var removeDuplicates = function (nums) {
    let left = 0;
    for (let right = 1; right < nums.length; right++)
        if (nums[left] != nums[right]) nums[++left] = nums[right];
    return left + 1;
};

File: bookOfProblems/leetcode.com/golang/two-pointers/26. Remove Duplicates from Sorted Array.go
func removeDuplicates(nums []int) int {
	// l - left - last unique element
	// r - right - iterator

	l := 0
	for r := 1; r < len(nums); r++ {
		if nums[r] != nums[l] {
			l++
			nums[l] = nums[r]
		}
	}

	return l + 1
}

////////////////////////////////////////////////////////////////////////////////

// Make use of 2 pointers, a fast and slow moving pointer. Initialise fast to 1 and slow to 0.
// While fast has not reached the end of array, check if the elements at index fast and slow are same, if yes, move fast forward by 1 place.
// If not, copy the non duplicate element into the position after slow.
// Once fast reaches the end of array, all unique elements will be stored inside the first "slow" number of elements in the array.

func removeDuplicates(nums []int) int {
	slow := 0
	fast := 1
	n := len(nums)

	for fast < n {
		if nums[fast] == nums[slow] {
			fast++
		} else {
			slow++
			nums[slow] = nums[fast]
		}
	}

	return slow + 1
}